main.go

```diff
@@ -68,10 +70,14 @@ var (
 	policiesMu   sync.RWMutex
 	policiesByID = map[string]user.Policy{}
 
-	mainRouter    *mux.Router
-	controlRouter *mux.Router
-	LE_MANAGER    letsencrypt.Manager
-	LE_FIRSTRUN   bool
+	mainRouter   *mux.Router
+	mainRouterMu sync.Mutex
+
+	controlRouter   *mux.Router
+	controlRouterMu sync.Mutex
```

Comment:
Also don't understand why these mutexes are needed. Do you have a race report that points directoy to the routers?

Reply:
these mutexes were added because of races 
i.e. if we remove all changes introduced to lock operations with `mainRouter`:
```
diff --git a/api_loader.go b/api_loader.go
index aab0b59..fd41b41 100644
--- a/api_loader.go
+++ b/api_loader.go
@@ -6,7 +6,6 @@ import (
 	"net/url"
 	"sort"
 	"strings"
-	"sync"
 	"sync/atomic"
 
 	"github.com/Sirupsen/logrus"
@@ -524,13 +523,11 @@ func loadGlobalApps() {
 	// we need to make a full copy of the slice, as loadApps will
 	// use in-place to sort the apis.
 	specs := getApiSpecs()
-	loadApps(specs, mainRouter, &mainRouterMu)
+	loadApps(specs, mainRouter)
 }
 
 // Create the individual API (app) specs based on live configurations and assign middleware
-func loadApps(specs []*APISpec, muxer *mux.Router, muxerMu *sync.Mutex) {
-	muxerMu.Lock()
-	defer muxerMu.Unlock()
+func loadApps(specs []*APISpec, muxer *mux.Router) {
 
 	hostname := config.Global.HostName
 	if hostname != "" {
diff --git a/api_test.go b/api_test.go
index ee2aca0..7b482ea 100644
--- a/api_test.go
+++ b/api_test.go
@@ -40,7 +40,7 @@ const apiTestDef = `{
 
 func loadSampleAPI(t *testing.T, def string) {
 	spec := createSpecTest(t, def)
-	loadApps([]*APISpec{spec}, discardMuxer, &discardMuxerMu)
+	loadApps([]*APISpec{spec}, discardMuxer)
 }
 
 type testAPIDefinition struct {
@@ -161,7 +161,7 @@ func TestApiHandlerPostDupPath(t *testing.T) {
 	apisMu.Lock()
 	apisByID = make(map[string]*APISpec)
 	apisMu.Unlock()
-	loadApps(specs(), discardMuxer, &discardMuxerMu)
+	loadApps(specs(), discardMuxer)
 
 	s2 = getApiSpec("2")
 	if want, got := "/v1-2", s2.Proxy.ListenPath; want != got {
@@ -177,9 +177,9 @@ func TestApiHandlerPostDupPath(t *testing.T) {
 	apisMu.Lock()
 	apisByID = make(map[string]*APISpec)
 	apisMu.Unlock()
-	loadApps(specs()[1:], discardMuxer, &discardMuxerMu)
-	loadApps(specs(), discardMuxer, &discardMuxerMu)
-	loadApps(specs(), discardMuxer, &discardMuxerMu)
+	loadApps(specs()[1:], discardMuxer)
+	loadApps(specs(), discardMuxer)
+	loadApps(specs(), discardMuxer)
 
 	s2 = getApiSpec("2")
 	if want, got := "/v1-2", s2.Proxy.ListenPath; want != got {
@@ -195,8 +195,8 @@ func TestApiHandlerPostDupPath(t *testing.T) {
 	apisMu.Lock()
 	apisByID = make(map[string]*APISpec)
 	apisMu.Unlock()
-	loadApps(specs(), discardMuxer, &discardMuxerMu)
-	loadApps(specs(), discardMuxer, &discardMuxerMu)
+	loadApps(specs(), discardMuxer)
+	loadApps(specs(), discardMuxer)
 
 	s2 = getApiSpec("2")
 	if want, got := "/v1-2", s2.Proxy.ListenPath; want != got {
@@ -670,9 +670,8 @@ func BenchmarkApiReload(b *testing.B) {
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
 		newMuxes := mux.NewRouter()
-		var newMuxesMu sync.Mutex
-		loadAPIEndpoints(newMuxes, &newMuxesMu)
-		loadApps(specs, newMuxes, &newMuxesMu)
+		loadAPIEndpoints(newMuxes)
+		loadApps(specs, newMuxes)
 	}
 }
 
@@ -747,8 +746,7 @@ func TestApiLoaderLongestPathFirst(t *testing.T) {
 	apisMu.Unlock()
 
 	mu := mux.NewRouter()
-	var muMutex sync.Mutex
-	loadApps(specs, mu, &muMutex)
+	loadApps(specs, mu)
 
 	for hp := range inputs {
 		rec := httptest.NewRecorder()
diff --git a/cert_test.go b/cert_test.go
index d4becc1..9f5d339 100644
--- a/cert_test.go
+++ b/cert_test.go
@@ -271,7 +271,7 @@ func TestGatewayControlAPIMutualTLS(t *testing.T) {
 		ln, _ := generateListener(0)
 		baseURL := "https://" + strings.Replace(ln.Addr().String(), "[::]", "127.0.0.1", -1)
 		listen(ln, nil, nil)
-		loadAPIEndpoints(mainRouter, &mainRouterMu)
+		loadAPIEndpoints(mainRouter)
 		defer func() {
 			ln.Close()
 			config.Global.HttpServerOptions.SSLCertificates = nil
diff --git a/gateway_test.go b/gateway_test.go
index 84d2a93..4e21f07 100644
--- a/gateway_test.go
+++ b/gateway_test.go
@@ -161,7 +161,7 @@ func TestMain(m *testing.M) {
 	afterConfSetup(&config.Global)
 	initialiseSystem(nil)
 	// Small part of start()
-	loadAPIEndpoints(mainRouter, &mainRouterMu)
+	loadAPIEndpoints(mainRouter)
 	if analytics.GeoIPDB == nil {
 		panic("GeoIPDB was not initialized")
 	}
@@ -841,7 +841,7 @@ func testHttp(t *testing.T, tests []tykHttpTest, separateControlPort bool) {
 		// This is emulate calling start()
 		// But this lines is the only thing needed for this tests
 		if config.Global.ControlAPIPort == 0 {
-			loadAPIEndpoints(mainRouter, &mainRouterMu)
+			loadAPIEndpoints(mainRouter)
 		}
 
 		if m.goagain {
diff --git a/main.go b/main.go
index 1a88571..a0c4cc6 100644
--- a/main.go
+++ b/main.go
@@ -70,8 +70,7 @@ var (
 	policiesMu   sync.RWMutex
 	policiesByID = map[string]user.Policy{}
 
-	mainRouter   *mux.Router
-	mainRouterMu sync.Mutex
+	mainRouter *mux.Router
 
 	controlRouter   *mux.Router
 	controlRouterMu sync.Mutex
@@ -101,12 +100,6 @@ func initGlobalEventsJSVM(spec *APISpec) {
 	GlobalEventsJSVM.Init(spec)
 }
 
-func resetMainRouter() {
-	mainRouterMu.Lock()
-	defer mainRouterMu.Unlock()
-	mainRouter = mux.NewRouter()
-}
-
 func resetControlRouter() {
 	controlRouterMu.Lock()
 	defer controlRouterMu.Unlock()
@@ -144,7 +137,8 @@ func apiSpecsLen() int {
 
 // Create all globals and init connection handlers
 func setupGlobals() {
-	resetMainRouter()
+	mainRouter = mux.NewRouter()
+
 	resetControlRouter()
 
 	if config.Global.EnableAnalytics && config.Global.Storage.Type != "redis" {
@@ -368,9 +362,7 @@ func controlAPICheckClientCertificate(certLevel string, next http.Handler) http.
 }
 
 // Set up default Tyk control API endpoints - these are global, so need to be added first
-func loadAPIEndpoints(muxer *mux.Router, muxerMu *sync.Mutex) {
-	muxerMu.Lock()
-	defer muxerMu.Unlock()
+func loadAPIEndpoints(muxer *mux.Router) {
 
 	hostname := config.Global.HostName
 	if config.Global.ControlAPIHostname != "" {
@@ -668,13 +660,13 @@ func doReload() {
 	log.WithFields(logrus.Fields{
 		"prefix": "main",
 	}).Info("Preparing new router")
-	resetMainRouter()
+	mainRouter = mux.NewRouter()
 	if config.Global.HttpServerOptions.OverrideDefaults {
 		mainRouter.SkipClean(config.Global.HttpServerOptions.SkipURLCleaning)
 	}
 
 	if config.Global.ControlAPIPort == 0 {
-		loadAPIEndpoints(mainRouter, &mainRouterMu)
+		loadAPIEndpoints(mainRouter)
 	}
 
 	loadGlobalApps()
@@ -1206,7 +1198,7 @@ func start(arguments map[string]interface{}) {
 		DefaultQuotaStore.Init(getGlobalStorageHandler("orgkey.", false))
 	}
 	if config.Global.ControlAPIPort == 0 {
-		loadAPIEndpoints(mainRouter, &mainRouterMu)
+		loadAPIEndpoints(mainRouter)
 	}
 
 	// Start listening for reload messages
@@ -1356,7 +1348,7 @@ func listen(l, controlListener net.Listener, err error) {
 			}
 
 			if config.Global.ControlAPIPort > 0 {
-				loadAPIEndpoints(controlRouter, &controlRouterMu)
+				loadAPIEndpoints(controlRouter)
 			}
 		}
 
@@ -1432,7 +1424,7 @@ func listen(l, controlListener net.Listener, err error) {
 			}
 
 			if config.Global.ControlAPIPort > 0 {
-				loadAPIEndpoints(controlRouter, &controlRouterMu)
+				loadAPIEndpoints(controlRouter)
 			}
 
 			if config.Global.UseDBAppConfigs {
diff --git a/mw_version_check_test.go b/mw_version_check_test.go
index 0e59091..ca7fc4f 100644
--- a/mw_version_check_test.go
+++ b/mw_version_check_test.go
@@ -7,7 +7,7 @@ import (
 
 func TestVersionMwExpiresHeader(t *testing.T) {
 	spec := createSpecTest(t, nonExpiringDef)
-	loadApps([]*APISpec{spec}, discardMuxer, &discardMuxerMu)
+	loadApps([]*APISpec{spec}, discardMuxer)
 
 	session := createNonThrottledSession()
 	spec.SessionManager.UpdateSession("1234xyz", session, 60)
diff --git a/oauth_manager_test.go b/oauth_manager_test.go
index 95a5ab1..9d7938e 100644
--- a/oauth_manager_test.go
+++ b/oauth_manager_test.go
@@ -74,7 +74,7 @@ const oauthDefinition = `{
 
 func getOAuthChain(spec *APISpec, muxer *mux.Router, muxerMu *sync.Mutex) {
 	// Ensure all the correct ahndlers are in place
-	loadAPIEndpoints(muxer, muxerMu)
+	loadAPIEndpoints(muxer)
 	manager := addOAuthHandlers(spec, muxer)
 
 	// add a test client
@@ -437,7 +437,7 @@ func TestOAuthAPIRefreshInvalidate(t *testing.T) {
 	tokenData := getToken(t)
 
 	spec := createSpecTest(t, oauthDefinition)
-	loadApps([]*APISpec{spec}, discardMuxer, &discardMuxerMu)
+	loadApps([]*APISpec{spec}, discardMuxer)
 	testMuxer := mux.NewRouter()
 	var testMuxerMu sync.Mutex
 
diff --git a/rpc_backup_handlers.go b/rpc_backup_handlers.go
index 8ebea5e..b9f8ca5 100644
--- a/rpc_backup_handlers.go
+++ b/rpc_backup_handlers.go
@@ -15,6 +15,7 @@ import (
 	"sync/atomic"
 
 	"github.com/Sirupsen/logrus"
+	"github.com/gorilla/mux"
 )
 
 const RPCKeyPrefix = "rpc:"
@@ -95,16 +96,16 @@ func doLoadWithBackup(specs []*APISpec) {
 	initGlobalEventsJSVM(nil)
 	log.Warning("[RPC Backup] --> Initialised JSVM")
 
-	resetMainRouter()
+	mainRouter = mux.NewRouter()
 
 	log.Warning("[RPC Backup] --> Set up routers")
 	log.Warning("[RPC Backup] --> Loading endpoints")
 
-	loadAPIEndpoints(mainRouter, &mainRouterMu)
+	loadAPIEndpoints(mainRouter)
 
 	log.Warning("[RPC Backup] --> Loading APIs")
 	currSpecs := getApiSpecs()
-	loadApps(currSpecs, mainRouter, &mainRouterMu)
+	loadApps(currSpecs, mainRouter)
 	log.Warning("[RPC Backup] --> API Load Done")
 
 	newServeMux := http.NewServeMux()
```
we get these races:
```
go test -race ./...
==================
WARNING: DATA RACE
Write at 0x000002537f00 by goroutine 189:
  github.com/TykTechnologies/tyk.doReload()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:663 +0x2cd
  github.com/TykTechnologies/tyk.reloadLoop()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:696 +0xe2

Previous write at 0x000002537f00 by goroutine 353:
  github.com/TykTechnologies/tyk.setupGlobals()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:140 +0x13f
  github.com/TykTechnologies/tyk.testHttp()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:840 +0xd85
  github.com/TykTechnologies/tyk.TestListener()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:948 +0xf4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c

Goroutine 189 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:169 +0x488
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 353 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:921 +0x206
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:192 +0x4c9
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
==================
WARNING: DATA RACE
Read at 0x00000253d908 by goroutine 189:
  github.com/TykTechnologies/tyk.doReload()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:664 +0x2f7
  github.com/TykTechnologies/tyk.reloadLoop()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:696 +0xe2

Previous write at 0x00000253d908 by goroutine 353:
  github.com/TykTechnologies/tyk.testHttp()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:819 +0xb92
  github.com/TykTechnologies/tyk.TestListener()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:948 +0xf4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c

Goroutine 189 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:169 +0x488
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 353 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:921 +0x206
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:192 +0x4c9
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
==================
WARNING: DATA RACE
Read at 0x00000253da40 by goroutine 189:
  github.com/TykTechnologies/tyk.doReload()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:668 +0x316
  github.com/TykTechnologies/tyk.reloadLoop()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:696 +0xe2

Previous write at 0x00000253da40 by goroutine 353:
  github.com/TykTechnologies/tyk.testHttp()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:816 +0xf5e
  github.com/TykTechnologies/tyk.TestListener()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:948 +0xf4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c

Goroutine 189 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:169 +0x488
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 353 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:921 +0x206
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:192 +0x4c9
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
==================
WARNING: DATA RACE
Read at 0x00c425a436a0 by goroutine 353:
  github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux.(*Router).PathPrefix()
      /Users/test/go/src/github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux/mux.go:213 +0x194
  github.com/TykTechnologies/tyk.loadAPIEndpoints()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:373 +0x1b1
  github.com/TykTechnologies/tyk.testHttp()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:844 +0xec6
  github.com/TykTechnologies/tyk.TestListener()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:948 +0xf4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c

Previous write at 0x00c425a436a0 by goroutine 189:
  github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux.(*Router).HandleFunc()
      /Users/test/go/src/github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux/mux.go:213 +0x220
  github.com/TykTechnologies/tyk.loadApps()
      /Users/test/go/src/github.com/TykTechnologies/tyk/api_loader.go:637 +0x1663
  github.com/TykTechnologies/tyk.loadGlobalApps()
      /Users/test/go/src/github.com/TykTechnologies/tyk/api_loader.go:526 +0x8e
  github.com/TykTechnologies/tyk.doReload()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:672 +0x32b
  github.com/TykTechnologies/tyk.reloadLoop()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:696 +0xe2

Goroutine 353 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:921 +0x206
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:192 +0x4c9
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 189 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:169 +0x488
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
==================
WARNING: DATA RACE
Read at 0x00c420a33f58 by goroutine 353:
  runtime.growslice()
      /usr/local/go/src/runtime/slice.go:82 +0x0
  github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux.(*Router).PathPrefix()
      /Users/test/go/src/github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux/mux.go:213 +0x300
  github.com/TykTechnologies/tyk.loadAPIEndpoints()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:373 +0x1b1
  github.com/TykTechnologies/tyk.testHttp()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:844 +0xec6
  github.com/TykTechnologies/tyk.TestListener()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:948 +0xf4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c

Previous write at 0x00c420a33f58 by goroutine 189:
  github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux.(*Router).HandleFunc()
      /Users/test/go/src/github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux/mux.go:213 +0x1ee
  github.com/TykTechnologies/tyk.loadApps()
      /Users/test/go/src/github.com/TykTechnologies/tyk/api_loader.go:637 +0x1663
  github.com/TykTechnologies/tyk.loadGlobalApps()
      /Users/test/go/src/github.com/TykTechnologies/tyk/api_loader.go:526 +0x8e
  github.com/TykTechnologies/tyk.doReload()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:672 +0x32b
  github.com/TykTechnologies/tyk.reloadLoop()
      /Users/test/go/src/github.com/TykTechnologies/tyk/main.go:696 +0xe2

Goroutine 353 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:921 +0x206
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:192 +0x4c9
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 189 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /Users/test/go/src/github.com/TykTechnologies/tyk/gateway_test.go:169 +0x488
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
--- FAIL: TestListener (0.63s)
	testing.go:699: race detected during execution of test
FAIL
FAIL	github.com/TykTechnologies/tyk	26.294s
?   	github.com/TykTechnologies/tyk/apidef	[no test files]
ok  	github.com/TykTechnologies/tyk/apidef/importer	1.041s
ok  	github.com/TykTechnologies/tyk/certs	5.526s
ok  	github.com/TykTechnologies/tyk/config	1.091s
?   	github.com/TykTechnologies/tyk/coprocess	[no test files]
ok  	github.com/TykTechnologies/tyk/lint	1.173s
?   	github.com/TykTechnologies/tyk/log	[no test files]
?   	github.com/TykTechnologies/tyk/storage	[no test files]
?   	github.com/TykTechnologies/tyk/user	[no test files]
```

The same problems if we don't lock access to `controlRouter`...